### DISCUSSION TITLE
How to make clangd not mark unicode identifiers valid before P1949 as errors (e.g. 'Unexpected character <U+...> clang(character_not_allowed))?

### DISCUSSION THREAD

**Discussion by yg-i** [clangd/clangd#2051] [Q&A] [answered]
I'm using clangd's VSCode extension.

Some unicode identifiers (just as an example: <U+1F361>, the emoji 🍡) was valid before P1949 ([1]; see discussion at [2]). 

Clang has decided to implement P1949, and as a result a lot of my code is broken. Fortunately I can still compile them using g++ or msvc. But I'd like to continue to use clangd as a language server.

Is there a way to indicate to clangd that it should ignore P1949?

Thanks in advance!

[1] https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p1949r7.html
[2] https://github.com/llvm/llvm-project/issues/54732

**Comment by HighCommander4** [comment]
Since clangd is implemented using clang's frontend, its behaviour is going to follow the behaviour of clang (the compiler).

So, either clang supports some option to get back the old behaviour (in which case, you can specify the option for clangd using something like https://clangd.llvm.org/config.html#compileflags), or it doesn't (in which case, clangd doesn't either).

**Reply by yg-i** [reply]
Thank you for the helpful pointer! I have tried:
```
	"clangd.fallbackFlags": [
		"-std=c++11"
	]
```
But still getting the same error. Which I don't quite understand, since as I understand it, P1949 is proposed for c++23 ("Adopt Unicode Annex 31 as part of C++ 23"; https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p1949r7.html)

I don't get it ... Did clang decide to _retroactively_ apply P1949 to earlier versions of C++? Maybe I should ask this question over there..

**Reply by HighCommander4** [reply, marked as answer]
Having reviewed the discussion at https://github.com/llvm/llvm-project/issues/54732, it looks like:

 * Yes, clang retroactively applied P1949 to earlier C++ versions. They did this because the committee marked the change as a "Defect Report", which means the previous behaviour is considered a bug in older standard versions.
 * Due to feedback received about a particular set of characters (mathematical operators and subscripts), clang downgraded its diagnostic about those characters in particular to a warning.
 * A more general option which continues to allow other characters was not added.

To me this seems fairly unfortunate as it's a breaking change.

That said, I don't think there's anything we can do about it at the clangd level. A change would have to come on the clang side. The discussion so far seems to have been specific to mathematical symbols (which is also why the mitigation that was rolled out was specific to those), so if you have a compelling use case for using other types of characters in identifiers, you may want to raise that for discussion in a clang issue.

**Comment by yg-i** [comment]
Thanks again for @HighCommander4's kind help! 

In case anyone is in the same situation, I solved the problem by simply replacing the clangd binary with the version in the 13.0.0 release (https://github.com/clangd/clangd/releases/tag/13.0.0). Now intellisense will no longer scream about our beloved symbols: 😀

<img src="https://github.com/clangd/clangd/assets/148152939/4bec484e-5314-474a-90b6-45a44dc925c0" width="500" />
